tests/test_typing.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Optional
+from uuid import UUID
+
+import falcon
+import falcon.asgi
+import falcon.testing
+
+
+@dataclass
+class RichContext:
+    userid: Optional[UUID] = None
+    role: str = 'anonymous'
+    comment: str = 'no comment'
+
+
+class FancyRequest(falcon.Request):
+    context_type = RichContext
+
+
+class FancyResponse(falcon.Response):
+    context_type = RichContext
+
+    # NOTE(vytas): the `type: ignore` exemption is currently required if it is
+    #   desirable to actually check typing of context attributes. See also:
+    #   https://falcon.readthedocs.io/en/latest/api/typing.html#known-limitations
```

Comment:
This is forward-deployed btw :shipit:. The link will only provide the relevant information after merging this.